test/CloudNative.CloudEvents.UnitTests/NewtonsoftJson/JsonEventFormatterTest.cs

```diff
@@ -2,76 +2,785 @@
+                ["id"] = "test-id",
+                ["source"] = SampleUriText,
+            };
+            byte[] bytes = Encoding.UTF8.GetBytes(obj.ToString());
+            var stream = new MemoryStream(bytes);
+            var formatter = new JsonEventFormatter();
+            var cloudEvent = await formatter.DecodeStructuredModeMessageAsync(stream, s_jsonCloudEventContentType, null);
+            Assert.Equal("test-type", cloudEvent.Type);
+            Assert.Equal("test-id", cloudEvent.Id);
+            Assert.Equal(SampleUri, cloudEvent.Source);
+        }
+
+        [Fact]
+        public void DecodeStructuredModeMessage_Minimal()
+        {
+            var obj = new JObject
+            {
+                ["specversion"] = "1.0",
+                ["type"] = "test-type",
+                ["id"] = "test-id",
+                ["source"] = SampleUriText,
+            };
+            var cloudEvent = DecodeStructuredModeMessage(obj);
+            Assert.Equal("test-type", cloudEvent.Type);
+            Assert.Equal("test-id", cloudEvent.Id);
+            Assert.Equal(SampleUri, cloudEvent.Source);
+        }
+
+        [Fact]
+        public void DecodeStructuredModeMessage_NoSpecVersion()
+        {
+            var obj = new JObject
+            {
+                ["type"] = "test-type",
+                ["id"] = "test-id",
+                ["source"] = SampleUriText,
+            };
+            Assert.Throws<ArgumentException>(() => DecodeStructuredModeMessage(obj));
+        }
+
+        [Fact]
+        public void DecodeStructuredModeMessage_UnknownSpecVersion()
+        {
+            var obj = new JObject
+            {
+                ["specversion"] = "0.5",
+                ["type"] = "test-type",
+                ["id"] = "test-id",
+                ["source"] = SampleUriText,
+            };
+            Assert.Throws<ArgumentException>(() => DecodeStructuredModeMessage(obj));
+        }
+
+        [Fact]
+        public void DecodeStructuredModeMessage_MissingRequiredAttributes()
+        {
+            var obj = new JObject
+            {
+                ["specversion"] = "1.0",
+                ["type"] = "test-type",
+                ["id"] = "test-id"
+                // Source is missing
+            };
+            // TODO: Arguably ArgumentException would be neater here.
+            Assert.Throws<InvalidOperationException>(() => DecodeStructuredModeMessage(obj));
+        }
+
+        [Fact]
+        public void DecodeStructuredModeMessage_SpecVersionNotString()
+        {
+            var obj = new JObject
+            {
+                ["specversion"] = 1,
+                ["type"] = "test-type",
+                ["id"] = "test-id",
+                ["source"] = SampleUriText,
+            };
+            Assert.Throws<ArgumentException>(() => DecodeStructuredModeMessage(obj));
+        }
+
+        [Fact(Skip = "Not yet implemented")]
+        public void DecodeStructuredModeMessage_TypeNotString()
```

Comment:
Although this somewhat contradicts my thoughts on the extension attribute topic, I think this should throw. It makes sense to be more strict for the reserved attributes. It might be worth thinking about having a parameter that can be set that allows validation to be skipped.

Reply:
Agreed for the first part. I think I'd want to wait until we get a feature request before allowing validation to be skipped, but I can see it *might* be useful.

Will implement in a follow-up PR (quite possibly tomorrow).